pkg/agent/controller/trafficcontrol/controller.go

```diff
@@ -0,0 +1,767 @@
+		if devConfig.VNI != nil {
+			vni = *devConfig.VNI
+		}
+		portName = strings.Join([]string{portNamePrefixVXLAN, genPortNameUDPTunnel(remoteIP, dstPort, vni)}, "-")
+	case device.GENEVE != nil:
+		devConfig := device.GENEVE
+		remoteIP := devConfig.RemoteIP
+		dstPort := defaultGENEVETunnelDestinationPort
+		if devConfig.DestinationPort != nil {
+			dstPort = *devConfig.DestinationPort
+		}
+		portName = strings.Join([]string{portNamePrefixGENEVE, genPortNameUDPTunnel(remoteIP, dstPort, 0)}, "-")
+	case device.GRE != nil:
+		portName = strings.Join([]string{portNamePrefixGRE, genPortNameGRETunnel(device.GRE)}, "-")
+	case device.ERSPAN != nil:
+		portName = strings.Join([]string{portNamePrefixERSPAN, genPortNameERSPANTunnel(device.ERSPAN)}, "-")
+	}
+	itf, ok := c.interfaceStore.GetInterfaceByName(portName)
+	if ok {
+		return uint32(itf.OFPort), true
+	} else {
+		return 0, false
+	}
+}
+
+func (c *Controller) DeleteTrafficControlPort(port uint32) error {
+	if itf, ok := c.interfaceStore.GetInterfaceByOFPort(port); ok {
+		defer c.interfaceStore.DeleteInterface(itf)
+		return c.ovsBridgeClient.DeletePort(itf.PortUUID)
+	} else {
+		return fmt.Errorf("interface %v not found", port)
+	}
+}
+
+func (c *Controller) syncTrafficControl(tcName string) error {
+	startTime := time.Now()
+	defer func() {
+		klog.V(2).InfoS("Finished syncing TrafficControl", "TrafficControl", tcName, "durationTime", time.Since(startTime))
+	}()
+
+	tc, err := c.trafficControlLister.Get(tcName)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			// If a TrafficControl is removed and the corresponding state doesn't exist, just return.
+			tcState, exist := c.getTCState(tcName)
+			if !exist {
+				return nil
+			}
+			// If a TrafficControl is removed but corresponding state exists, the state is not needed and remove it.
+			if err = c.uninstallTrafficControl(tcName, tcState); err != nil {
+				return err
+			}
+			return nil
+		}
+		return err
+	}
+
+	tcState, exist := c.getTCState(tcName)
+	if !exist {
+		// If corresponding state of a TrafficControl doesn't exist, create it.
+		tcState = c.createTCState(tc)
+	}
+
+	var returnPort, targetPort uint32
+	if tc.Spec.ReturnPort != nil {
+		returnPort, exist = c.getTrafficControlPort(tc.Spec.ReturnPort)
+		if !exist {
+			// If the return port doesn't exist, create the port.
+			if returnPort, err = c.createTrafficControlPort(tc.Spec.ReturnPort); err != nil {
+				return err
+			}
+		}
+		if err = c.ofClient.InstallTrafficControlReturnPortFlow(returnPort); err != nil {
+			return err
+		}
+		// If the return port exists and is changed:
+		//   - Install the new return flow.
+		//   - Uninstall the stale return flow.
+		//   - Clean the stale OF port.
+		if tcState.returnPort != 0 && returnPort != tcState.returnPort {
+			if err = c.ofClient.UninstallTrafficControlReturnPortFlow(tcState.returnPort); err != nil {
+				return err
+			}
+			if err = c.DeleteTrafficControlPort(tcState.returnPort); err != nil {
```

Comment:
Check if it's still used by any TC before actually deleting it.

Reply:
Done